File: scripts/probe_fuyao.py
"""Probe Fuyao access without printing keys or source values."""

from __future__ import annotations

import json
import re
import sys
import time
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlencode, urlparse
from urllib.request import Request

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "src"))

from diagnosis.net import load_env, open_direct  # noqa: E402


def probe(base: str, key: str, name: str, path: str, params: dict) -> dict:
    started = time.monotonic()
    request = Request(f"{base}{path}?{urlencode(params)}", headers={"X-api-key": key})
    try:
        with open_direct(request, timeout=20) as response:
            http_status = response.status
            body = json.load(response)
        data = body.get("data") or {}
        items = data.get("item")
        abilities = data.get("abilities")
        count = len(items) if isinstance(items, list) else sum(
            len(ability.get("indicators", [])) for ability in abilities if isinstance(ability, dict)
        ) if isinstance(abilities, list) else None
        first = items[0] if isinstance(items, list) and items else None
        return {
            "name": name, "http_status": http_status, "business_code": body.get("code"),
            "status": "http_error" if http_status != 200 else "api_error" if body.get("code") != 0 else
            "empty" if not count else "ok",
            "message": str(body.get("message") or "").replace(key, "[redacted]")[:160],
            "item_count": count,
            "data_shape": "item" if isinstance(items, list) else "abilities" if isinstance(abilities, list) else "other",
            "data_timestamp_ms": data.get("timestamp"), "report_period": data.get("report"),
            "first_period_end_ms": first.get("period_end_ms") if isinstance(first, dict) else None,
            "first_item_fields": list(first)[:40] if isinstance(first, dict) else [],
            "elapsed_ms": round((time.monotonic() - started) * 1000),
        }
    except Exception as error:
        return {"name": name, "status": "request_failed", "error_type": type(error).__name__,
                "elapsed_ms": round((time.monotonic() - started) * 1000)}


def main() -> int:
    env = load_env(ROOT / ".env")
    key = env.get("FUYAO_API_KEY", "")
    base = urlparse(env.get("FUYAO_BASE_URL") or "https://fuyao.aicubes.cn")
    stock = sys.argv[1] if len(sys.argv) > 1 else "600519.SH"
    if not key or base.scheme != "https" or base.netloc != "fuyao.aicubes.cn" or not re.fullmatch(r"\d{6}\.(SH|SZ|BJ)", stock):
        print("Missing key, unofficial Fuyao URL, or invalid A-share code", file=sys.stderr)
        return 2
    end = round(time.time() * 1000)
    start = end - 60 * 24 * 60 * 60 * 1000
    requests = [
        ("snapshot", "/api/a-share/prices/snapshot", {"thscodes": stock}),
        ("historical", "/api/a-share/prices/historical", {"thscode": stock, "interval": "1d", "adjust": "forward", "start": start, "end": end}),
        ("income", "/api/a-share/financials/income-statements", {"thscode": stock, "period": "quarterly", "limit": 4}),
        ("balance", "/api/a-share/financials/balance-sheets", {"thscode": stock, "period": "quarterly", "limit": 4}),
        ("cash_flow", "/api/a-share/financials/cash-flow-statements", {"thscode": stock, "period": "quarterly", "limit": 4}),
        ("indicators", "/api/a-share/financials/indicators", {"thscode": stock, "report": "2026-2"}),
        ("valuation", "/api/a-share/valuations/snapshot", {"thscodes": stock}),
    ]
    with ThreadPoolExecutor(max_workers=len(requests)) as pool:
        origin = f"{base.scheme}://{base.netloc}"
        results = list(pool.map(lambda item: probe(origin, key, *item), requests))
    print(json.dumps({"checked_at": datetime.now(timezone.utc).isoformat(), "sample_stock": stock,
                      "results": results}, ensure_ascii=False, indent=2))
    return 1 if all(item["status"] != "ok" for item in results) else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/probe-fuyao.mjs
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

function loadEnv(path) {
  const values = {};
  for (const line of readFileSync(path, 'utf8').split(/\r?\n/)) {
    const match = line.match(/^([A-Z][A-Z0-9_]*)=(.*)$/);
    if (match) values[match[1]] = match[2];
  }
  return values;
}

const env = loadEnv(resolve(process.cwd(), '.env'));
const key = env.FUYAO_API_KEY;
const base = env.FUYAO_BASE_URL || 'https://fuyao.aicubes.cn';
const baseUrl = new URL(base);
if (baseUrl.protocol !== 'https:' || baseUrl.hostname !== 'fuyao.aicubes.cn') {
  console.error('FUYAO_BASE_URL must be the official Fuyao HTTPS endpoint');
  process.exit(2);
}
const stock = process.argv[2] || '600519.SH';
if (!key) {
  console.error('FUYAO_API_KEY is missing from the local .env');
  process.exit(2);
}
if (!/^\d{6}\.(SH|SZ|BJ)$/.test(stock)) {
  console.error('Stock must be a full A-share thscode, such as 600519.SH');
  process.exit(2);
}

const end = Date.now();
const start = end - 60 * 24 * 60 * 60 * 1000;
const probes = [
  ['snapshot', '/api/a-share/prices/snapshot', { thscodes: stock }],
  ['historical', '/api/a-share/prices/historical', { thscode: stock, interval: '1d', adjust: 'forward', start, end }],
  ['income', '/api/a-share/financials/income-statements', { thscode: stock, period: 'quarterly', limit: 4 }],
  ['balance', '/api/a-share/financials/balance-sheets', { thscode: stock, period: 'quarterly', limit: 4 }],
  ['cash_flow', '/api/a-share/financials/cash-flow-statements', { thscode: stock, period: 'quarterly', limit: 4 }],
  ['indicators', '/api/a-share/financials/indicators', { thscode: stock, report: '2026-2' }],
  ['valuation', '/api/a-share/valuations/snapshot', { thscodes: stock }],
];

async function probe([name, path, params]) {
  const url = new URL(path, base);
  for (const [param, value] of Object.entries(params)) url.searchParams.set(param, String(value));
  const started = Date.now();
  try {
    const response = await fetch(url, {
      headers: { 'X-api-key': key },
      signal: AbortSignal.timeout(20000),
    });
    const body = await response.json();
    const items = body?.data?.item;
    const abilities = body?.data?.abilities;
    const count = Array.isArray(items)
      ? items.length
      : Array.isArray(abilities)
        ? abilities.reduce((total, ability) => total + (Array.isArray(ability.indicators) ? ability.indicators.length : 0), 0)
        : null;
    const first = Array.isArray(items) ? items[0] : null;
    return {
      name,
      http_status: response.status,
      business_code: body?.code ?? null,
      status: !response.ok ? 'http_error' : body?.code !== 0 ? 'api_error' : !count ? 'empty' : 'ok',
      message: String(body?.message ?? '').replaceAll(key, '[redacted]').slice(0, 160),
      item_count: count,
      data_shape: Array.isArray(items) ? 'item' : Array.isArray(abilities) ? 'abilities' : 'other',
      data_timestamp_ms: body?.data?.timestamp ?? null,
      report_period: body?.data?.report ?? null,
      first_period_end_ms: first?.period_end_ms ?? null,
      first_item_fields: first && typeof first === 'object' ? Object.keys(first).slice(0, 40) : [],
      elapsed_ms: Date.now() - started,
    };
  } catch (error) {
    return { name, status: 'request_failed', error_type: error?.name || 'Error', elapsed_ms: Date.now() - started };
  }
}

const results = await Promise.all(probes.map(probe));
console.log(JSON.stringify({ checked_at: new Date().toISOString(), sample_stock: stock, results }, null, 2));
if (results.every((result) => result.status !== 'ok')) process.exitCode = 1;
